File: src/util/private-dir.ts
import { chmodSync, lstatSync, mkdirSync } from "node:fs";
import { dirname, join } from "node:path";

/**
 * Component-wise private-directory guarantee for Beam-owned local storage.
 *
 * Beam's local storage holds full workspace mirrors, transcripts, and
 * ownership tokens; the default umask (022) would leave every created
 * parent traversable by any local user. Starting at `root` — a Beam-owned
 * boundary such as BEAM_DIR — and descending through `segments`, every
 * component is proven to be a real directory owned by this process and is
 * closed to group/other:
 *
 * - missing        -> created 0700 (a mkdir mode is masked by the umask,
 *                     but the umask only CLEARS bits and 0700 has no
 *                     group/other bits to clear — so 0700 is umask-immune)
 * - symlink        -> refuse: a replaced component would silently redirect
 *                     every later stage/receipt write outside Beam's storage
 * - non-directory  -> refuse
 * - foreign owner  -> refuse (where the platform reports ownership): Beam
 *                     never chmods or writes beneath another user's object
 * - owned but open -> chmod 0700: a dir Beam created before this hardening
 *                     (or under a broader umask) is retro-tightened
 *
 * After each component the entry is re-read and must be a real directory
 * with no group/other bits — a failed or raced tightening refuses instead
 * of proceeding into disclosable storage. Ancestors ABOVE `root` are
 * outside Beam's ownership boundary and are never chmodded; they are
 * created plainly when missing (disclosure protection starts at `root`,
 * whose 0700 blocks traversal into everything below).
 *
 * Returns the fully joined, verified path.
 */
export function ensurePrivateBeamDir(root: string, ...segments: string[]): string {
  mkdirSync(dirname(root), { recursive: true });
  let p = root;
  securePrivateComponent(p);
  for (const seg of segments) {
    p = join(p, seg);
    securePrivateComponent(p);
  }
  return p;
}

/** Read-only counterpart used before consuming an existing private receipt. */
export function assertPrivateBeamDir(root: string, ...segments: string[]): string {
  let path = root;
  assertPrivateComponent(path);
  for (const segment of segments) {
    path = join(path, segment);
    assertPrivateComponent(path);
  }
  return path;
}

function securePrivateComponent(p: string): void {
  let st = lstatSync(p, { throwIfNoEntry: false });
  if (st === undefined) {
    try {
      mkdirSync(p, { mode: 0o700 });
    } catch (err) {
      // Only EEXIST is expected here: a lost same-user create race, where
      // the re-inspection below decides what the winner left behind. Any
      // other fault (EACCES, ENOSPC, ENOTDIR, ...) is real and surfaces.
      if ((err as NodeJS.ErrnoException).code !== "EEXIST") throw err;
    }
    st = lstatSync(p, { throwIfNoEntry: false });
    if (st === undefined) throw new Error(`beam: failed to create private directory ${p}`);
  }
  if (st.isSymbolicLink()) {
    throw new Error(
      `beam: ${p} is a symlink — Beam's private storage must be a real directory it owns; ` +
        `refusing to follow it. Move the link aside and retry`,
    );
  }
  if (!st.isDirectory()) {
    throw new Error(`beam: ${p} exists but is not a directory — move it aside and retry`);
  }
  const uid = typeof process.getuid === "function" ? process.getuid() : undefined;
  if (uid !== undefined && st.uid !== uid) {
    throw new Error(
      `beam: ${p} is owned by uid ${st.uid}, not this process (uid ${uid}) — refusing to use ` +
        `foreign storage for private return data. Move it aside and retry`,
    );
  }
  if ((st.mode & 0o077) !== 0) chmodSync(p, 0o700);
  assertPrivateComponent(p);
}

function assertPrivateComponent(path: string): void {
  const st = lstatSync(path, { throwIfNoEntry: false });
  if (st === undefined) {
    throw new Error(`beam: private directory ${path} is missing`);
  }
  const uid = typeof process.getuid === "function" ? process.getuid() : undefined;
  if (uid !== undefined && st.uid !== uid) {
    throw new Error(
      `beam: ${path} is owned by uid ${st.uid}, not this process (uid ${uid}) — ` +
        `refusing to read foreign private data`,
    );
  }
  if (st.isSymbolicLink() || !st.isDirectory() || (st.mode & 0o077) !== 0) {
    throw new Error(
      `beam: ${path} must be a real private directory owned by this process ` +
        `(mode ${(st.mode & 0o7777).toString(8)})`,
    );
  }
}

File: rust/src/util/private_dir.rs
//! Component-wise protection for Beam-owned local storage. Each component is
//! opened with `O_NOFOLLOW`, proven to be owned by this uid, tightened to
//! 0700 through the opened descriptor, then re-proven at the pathname.

use std::fmt::{Display, Formatter};
use std::fs::{self, DirBuilder, File, Permissions};
use std::os::unix::fs::{DirBuilderExt, MetadataExt, PermissionsExt};
use std::path::{Path, PathBuf};

use rustix::fs::{Mode, OFlags};
use rustix::process::getuid;

#[derive(Debug)]
pub struct PrivateDirError {
    message: String,
}

impl Display for PrivateDirError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl std::error::Error for PrivateDirError {}

/// Create or tighten `root` and every descendant segment. Ancestors above
/// `root` are created plainly and never chmodded; privacy starts at `root`.
pub fn ensure_private_beam_dir(root: &Path, segments: &[&str]) -> Result<PathBuf, PrivateDirError> {
    if root.as_os_str().is_empty() {
        return Err(PrivateDirError {
            message: "beam: private directory root must not be empty".to_owned(),
        });
    }
    if let Some(parent) = root.parent()
        && !parent.as_os_str().is_empty()
    {
        fs::create_dir_all(parent).map_err(|source| PrivateDirError {
            message: format!(
                "beam: could not create parent {} for private storage: {source}",
                parent.display()
            ),
        })?;
    }
    let mut path = root.to_path_buf();
    secure_private_component(&path)?;
    for segment in segments {
        validate_private_segment(segment)?;
        path.push(segment);
        secure_private_component(&path)?;
    }
    Ok(path)
}

fn validate_private_segment(segment: &str) -> Result<(), PrivateDirError> {
    if segment.is_empty() {
        return Err(PrivateDirError {
            message: "beam: private directory segment must not be empty".to_owned(),
        });
    }
    if segment == "." {
        return Err(PrivateDirError {
            message: "beam: private directory segment must not be .".to_owned(),
        });
    }
    if segment == ".." {
        return Err(PrivateDirError {
            message: "beam: private directory segment must not be ..".to_owned(),
        });
    }
    if segment.contains('/') {
        return Err(PrivateDirError {
            message: format!(
                "beam: private directory segment {segment:?} must be one path component"
            ),
        });
    }
    Ok(())
}

fn secure_private_component(path: &Path) -> Result<(), PrivateDirError> {
    let metadata = create_or_inspect_component(path)?;
    if metadata.file_type().is_symlink() {
        return Err(PrivateDirError {
            message: format!(
                "beam: {} is a symlink — Beam's private storage must be a real directory it owns; \
                 refusing to follow it. Move the link aside and retry",
                path.display()
            ),
        });
    }
    if !metadata.is_dir() {
        return Err(PrivateDirError {
            message: format!(
                "beam: {} exists but is not a directory — move it aside and retry",
                path.display()
            ),
        });
    }
    let directory = open_private_component(path)?;
    let opened = tighten_private_component(&directory, path)?;
    verify_private_component(path, &opened)
}

fn create_or_inspect_component(path: &Path) -> Result<std::fs::Metadata, PrivateDirError> {
    match fs::symlink_metadata(path) {
        Ok(metadata) => Ok(metadata),
        Err(source) if source.kind() == std::io::ErrorKind::NotFound => {
            let mut builder = DirBuilder::new();
            builder.mode(0o700);
            match builder.create(path) {
                Ok(()) => {}
                Err(source) if source.kind() == std::io::ErrorKind::AlreadyExists => {}
                Err(source) => {
                    return Err(PrivateDirError {
                        message: format!(
                            "beam: could not create private directory {}: {source}",
                            path.display()
                        ),
                    });
                }
            }
            fs::symlink_metadata(path).map_err(|source| PrivateDirError {
                message: format!(
                    "beam: failed to inspect created private directory {}: {source}",
                    path.display()
                ),
            })
        }
        Err(source) => Err(PrivateDirError {
            message: format!(
                "beam: could not inspect private directory {}: {source}",
                path.display()
            ),
        }),
    }
}

fn open_private_component(path: &Path) -> Result<File, PrivateDirError> {
    let fd = rustix::fs::open(
        path,
        OFlags::RDONLY | OFlags::CLOEXEC | OFlags::DIRECTORY | OFlags::NOFOLLOW,
        Mode::empty(),
    )
    .map_err(|source| PrivateDirError {
        message: format!(
            "beam: could not safely open private directory {}: {source}",
            path.display()
        ),
    })?;
    Ok(File::from(fd))
}

fn tighten_private_component(
    directory: &File,
    path: &Path,
) -> Result<std::fs::Metadata, PrivateDirError> {
    let mut metadata = directory.metadata().map_err(|source| PrivateDirError {
        message: format!(
            "beam: could not inspect opened private directory {}: {source}",
            path.display()
        ),
    })?;
    let uid = getuid().as_raw();
    if metadata.uid() != uid {
        return Err(PrivateDirError {
            message: format!(
                "beam: {} is owned by uid {}, not this process (uid {uid}) — refusing to use \
                 foreign storage for private return data. Move it aside and retry",
                path.display(),
                metadata.uid()
            ),
        });
    }
    if metadata.mode() & 0o077 != 0 {
        directory
            .set_permissions(Permissions::from_mode(0o700))
            .map_err(|source| PrivateDirError {
                message: format!(
                    "beam: could not tighten private directory {} to mode 700: {source}",
                    path.display()
                ),
            })?;
        metadata = directory.metadata().map_err(|source| PrivateDirError {
            message: format!(
                "beam: could not recheck private directory {}: {source}",
                path.display()
            ),
        })?;
    }
    Ok(metadata)
}

fn verify_private_component(
    path: &Path,
    opened: &std::fs::Metadata,
) -> Result<(), PrivateDirError> {
    let current = fs::symlink_metadata(path).map_err(|source| PrivateDirError {
        message: format!(
            "beam: private directory {} changed during verification: {source}",
            path.display()
        ),
    })?;
    if current.file_type().is_symlink() {
        return Err(PrivateDirError {
            message: format!(
                "beam: private directory {} became a symlink during verification — retry",
                path.display()
            ),
        });
    }
    if !current.is_dir() {
        return Err(PrivateDirError {
            message: format!(
                "beam: private directory {} stopped being a directory during verification — retry",
                path.display()
            ),
        });
    }
    if current.dev() != opened.dev() {
        return Err(PrivateDirError {
            message: format!(
                "beam: private directory {} changed device during verification — retry",
                path.display()
            ),
        });
    }
    if current.ino() != opened.ino() {
        return Err(PrivateDirError {
            message: format!(
                "beam: private directory {} changed inode during verification — retry",
                path.display()
            ),
        });
    }
    if current.mode() & 0o077 != 0 {
        return Err(PrivateDirError {
            message: format!(
                "beam: {} must be private mode 700 (found mode {:o})",
                path.display(),
                current.mode() & 0o7777
            ),
        });
    }
    Ok(())
}
